pkg/endpoints/pipeline_test.go

```diff
@@ -14,1408 +14,1039 @@ limitations under the License.
+		if err != nil {
+			t.Errorf("Error creating pipeline '%s': %v\n", pipeline.Name, err)
+		}
 	}
-	_, err = r.K8sClient.CoreV1().Pods("ns1").Create(&Pod3)
+	// Detected pipeline resources always attempt to be created
+	// All Resources names should be different between pipelineRuns
+	manualPipelineRuns := []ManualPipelineRun{
+		ManualPipelineRun{
+			PIPELINENAME: pipelineRefName,
+		},
+		ManualPipelineRun{
+			PIPELINENAME:    pipelineRefName,
+			GITRESOURCENAME: "git-source1",
+			GITCOMMIT:       "12345",
+			REPOURL:         "http://github.com/testorg/testrepo",
+		},
+		ManualPipelineRun{
+			PIPELINENAME:      pipelineRefName,
+			IMAGERESOURCENAME: "image-source1",
+			GITCOMMIT:         "12345",
+			REPONAME:          "testreponame",
+		},
+		ManualPipelineRun{
+			PIPELINENAME:      pipelineRefName,
+			IMAGERESOURCENAME: "image-source2",
+			GITRESOURCENAME:   "git-source2",
+			GITCOMMIT:         "12345",
+			REPONAME:          "testreponame",
+			REPOURL:           "http://github.com/testorg/testrepo",
+		},
+	}
+	// Test createPipelineRun function: bad request
+	httpRequestBody := strings.NewReader(`{"fielddoesnotexist": "pipeline",}`)
+	httpReq := dummyHTTPRequest("POST", fmt.Sprintf("%s/v1/namespaces/%s/pipelineruns", server.URL, namespace), httpRequestBody)
+	response, err := http.DefaultClient.Do(httpReq)
 	if err != nil {
-		t.Errorf("test error: %s", err)
+		t.Errorf("Error getting createPipelineRun response for pipelineRun '%s': %v\n", pipelineRefName, err)
+	} else {
+		expectedStatusCode := 400
+		if response.StatusCode != expectedStatusCode {
+			t.Errorf("Response code %d did not equal expected %d\n", response.StatusCode, expectedStatusCode)
+		}
 	}
-	_, err = r.K8sClient.CoreV1().Pods("ns1").Create(&Pod4)
+	// Test createPipelinerun function: non-existant pipeline reference
+	httpRequestBody = strings.NewReader(`{"pipelinename": "Pipelinedoesnotexist"}`)
+	httpReq = dummyHTTPRequest("POST", fmt.Sprintf("%s/v1/namespaces/%s/pipelineruns", server.URL, namespace), httpRequestBody)
+	response, err = http.DefaultClient.Do(httpReq)
 	if err != nil {
-		t.Errorf("test error: %s", err)
+		t.Errorf("Error getting createPipelineRun response for pipelineRun '%s': %v\n", pipelineRefName, err)
+	} else {
+		expectedStatusCode := 412
+		if response.StatusCode != expectedStatusCode {
+			t.Errorf("Response code %d did not equal expected %d\n", response.StatusCode, expectedStatusCode)
+		}
 	}
-
-	// Test getAllPipelineRuns function
-	httpReq := dummyHTTPRequest("GET", "http://wwww.dummy.com:8383/v1/namespaces/ns1/taskrunlogs/TaskRun1", nil)
-	req := dummyRestfulRequest(httpReq, "ns1", "TaskRun1")
-	httpWriter := httptest.NewRecorder()
-	resp := dummyRestfulResponse(httpWriter)
-
-	r.getTaskRunLog(req, resp)
-	var taskRunLog TaskRunLog
-	if err := json.NewDecoder(httpWriter.Body).Decode(&taskRunLog); err != nil {
-		t.Error(err)
+	// Test createPipelineRun function: good requests with/without pipelineResources
```

Comment:
do tests with and without pipelineResources belong in separate test methods?

Reply:
Most of the ~300-400 lines cut down (not that we have to) were due to the setup being the same since the same method is being tested. In this way, you only have to look at one area rather that making sure all your changes are in the distributed tests for the single endpoint. I can break it back to how it was before if you prefer, but personally I find this way a bit easier. I understand the createPipelineRun good request for pipelineResources was a bit bloated, but besides that we didn't break up getOne/getAll in other locations so I went one step further.

Reply:
no fair enough, let's take it as-is.